src/components/contactForm/ContactForm.svelte

```diff
@@ -1,39 +1,133 @@
+		required: boolean;
+	};
+
+	let nameFormItem: FormItem = {
+		formType: FormType.INPUT,
+		fieldName: 'Name',
+		itemValue: '',
+		isInputValid: false,
+		required: true
+	};
+
+	let emailFormItem: FormItem = {
+		formType: FormType.INPUT,
+		fieldName: 'Email',
+		itemValue: '',
+		isInputValid: false,
+		required: true
+	};
+
+	let messageFormItem: FormItem = {
+		formType: FormType.TEXTBOX,
+		fieldName: 'Message',
+		itemValue: '',
+		isInputValid: false,
+		required: false
+	};
+
+	let formItems = [nameFormItem, emailFormItem, messageFormItem];
+
 	let sent = false;
+	let isFormValid = true;
+	let errorMessage = '';
+	let isUserTyping: boolean;
+
+	const validateInput = (value: string) => {
+		return value !== undefined && value.length > 0;
+	};
+
+	const generateErrorMessage = () => {
+		let invalidItems = formItems.filter((item) => !item.isInputValid && item.required);
+		let invalidNames = invalidItems.map((item) => item.fieldName).join(' and ');
+
+		errorMessage = `Invalid or no input for ${invalidNames}`;
+	};
+
+	const validateForm = () => {
+		// accumulates the result of validateFormItem for each required formItem
+		isFormValid = formItems.reduce((acc, formItem) => {
+			if (formItem.required) {
+				formItem.isInputValid = validateInput(formItem.itemValue);
+				return acc && formItem.isInputValid;
+			}
+			return acc;
+		}, true);
+	};
+
+	const sendFormData = async () => {
+		if (!sent) {
+			fetch('/api/contact', {
+				method: 'post',
+				body: JSON.stringify({
+					data: {
+						name: nameFormItem.itemValue,
+						email: emailFormItem.itemValue,
+						message: messageFormItem.itemValue,
+						page
+					}
+				})
+			});
+			sent = true;
+		}
+	};
 
 	async function submit(e: Event): Promise<void> {
-		if (sent) {
-			return;
+		isUserTyping = !isUserTyping;
+		validateForm();
+
+		if (!isFormValid) {
+			return generateErrorMessage();
+		}
+
+		if (!sent) {
+			e.preventDefault();
+			await sendFormData();
 		}
-		e.preventDefault();
-		fetch('/api/contact', {
-			method: 'post',
-			body: JSON.stringify({
-				data: {
-					name,
-					email,
-					message,
-					page
-				}
-			})
-		});
-		sent = true;
 	}
+
+	const handleInput = (e: CustomEvent) => {
+		isUserTyping = e.detail.isUserTyping;
+	};
 </script>
 
 <form on:submit={submit}>
 	<h3>Contact us</h3>
-	<TextInput fieldName="Name" bind:value={name} />
-	<TextInput fieldName="Email" bind:value={email} />
 
-	<TextBoxInput fieldName="Message" bind:value={message} />
+	{#each formItems as formItem}
+		{#if formItem.formType === FormType.INPUT}
+			<TextInput
+				on:input={handleInput}
+				required={formItem.required}
+				fieldName={formItem.fieldName}
+				bind:value={formItem.itemValue}
+			/>
+		{/if}
+
+		{#if formItem.formType === FormType.TEXTBOX}
+			<TextBoxInput fieldName="Message" bind:value={formItem.itemValue} />
+		{/if}
+	{/each}
+
+	{#if !isUserTyping && !isFormValid}
+		<p class={!isFormValid ? 'error' : 'error__hide'}>{errorMessage}</p>
```

Comment:
I'd prefer if we were to show the error per field, close to the field.  Maybe something more like this?
![Error messages](https://user-images.githubusercontent.com/1701274/216802855-6c2b9421-2092-408b-8a7e-a708b934fd1f.gif)